core/src/main/java/com/linecorp/armeria/client/circuitbreaker/AbstractCircuitBreakerClient.java

```diff
@@ -115,13 +125,25 @@ final CircuitBreakerRule fromRuleWithContent() {
         return fromRuleWithContent;
     }
 
+    final CircuitBreakerClientHandler handler() {
+        return handler;
+    }
+
     @Override
     public final O execute(ClientRequestContext ctx, I req) throws Exception {
-        final CircuitBreakerCallback callback = handler.tryRequest(ctx, req);
-        if (callback == null) {
-            return unwrap().execute(ctx, req);
+        try {
+            final CircuitBreakerCallback callback = handler.tryRequest(ctx, req);
+            if (callback == null) {
+                return unwrap().execute(ctx, req);
+            }
+            return doExecute(ctx, req, callback);
+        } catch (Exception ex) {
+            if (handler().isCircuitBreakerException(ex) && fallback != null) {
```

Comment:
null checking would be cheaper than the combination of method invocations and type checking. 
```suggestion
            if (fallback != null && handler().isCircuitBreakerException(ex)) {
```